ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveInsertExchange4JoinRule.java

```diff
@@ -56,20 +57,28 @@
  */
 public class HiveInsertExchange4JoinRule extends RelOptRule {
 
-  protected static transient final Logger LOG = LoggerFactory
-      .getLogger(HiveInsertExchange4JoinRule.class);
+  protected static final Logger LOG = LoggerFactory.getLogger(HiveInsertExchange4JoinRule.class);
 
   /** Rule that creates Exchange operators under a MultiJoin operator. */
-  public static final HiveInsertExchange4JoinRule EXCHANGE_BELOW_MULTIJOIN =
-      new HiveInsertExchange4JoinRule(HiveMultiJoin.class);
+  public static HiveInsertExchange4JoinRule exchangeBelowMultiJoin(
+          RelFieldCollation.NullDirection defaultAscNullDirection) {
+    return new HiveInsertExchange4JoinRule(HiveMultiJoin.class, defaultAscNullDirection);
+  }
+
 
   /** Rule that creates Exchange operators under a Join operator. */
-  public static final HiveInsertExchange4JoinRule EXCHANGE_BELOW_JOIN =
-      new HiveInsertExchange4JoinRule(Join.class);
+  public static HiveInsertExchange4JoinRule exchangeBelowJoin(
+          RelFieldCollation.NullDirection defaultAscNullDirection) {
+    return new HiveInsertExchange4JoinRule(Join.class, defaultAscNullDirection);
+  }
+
+  private final RelFieldCollation.NullDirection defaultAscNullDirection;
```

Comment:
I don't see a big advantage in introducing the static factories. Since the constructor is public we could just keep that one. If we want to make the rule creation shorter then we could have a static factory with a shorter name (e.g., `of(clazz,nullDirection)`)

Reply:
Removed the factory methods. Using constructor in CalcitePlanner